tools/amgh/main.go

```diff
@@ -0,0 +1,128 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"path/filepath"
+
+	"github.com/google/go-github/v26/github"
+	"github.com/prometheus/alertmanager/notify"
+	"golang.org/x/oauth2"
+	"gopkg.in/alecthomas/kingpin.v2"
+)
+
+type ghWebhookRecieverConfig struct {
+	authfile     string
+	defaultOwner string
+	defaultRepo  string
+	portNo       string
+}
+
+type ghWebhookReciever struct {
+	ghClient *github.Client
+	cfg      ghWebhookRecieverConfig
+}
+
+type ghWebhookHandler struct {
+	client *ghWebhookReciever
+}
+
+func (hl ghWebhookHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if r.Method != http.MethodPost {
+		log.Printf("unsupported request method: %v: %v", r.Method, r.RemoteAddr)
+		w.WriteHeader(http.StatusMethodNotAllowed)
+		return
+	}
+
+	msg := &notify.WebhookMessage{}
+	ctx := r.Context()
+
+	// Decode the webhook request.
+	err := json.NewDecoder(r.Body).Decode(&msg)
+	if err != nil {
+		log.Println("failed to decode webhook data")
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	// Handle the webhook message.
+	log.Printf("handling alert: %v", alertID(msg))
+	if err := hl.client.processAlert(ctx, msg); err != nil {
+		log.Printf("failed to handle alert: %v: %v", alertID(msg), err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	log.Printf("completed alert: %v", alertID(msg))
+	w.WriteHeader(http.StatusOK)
+}
+
+func newGhWebhookReciever(cfg ghWebhookRecieverConfig) (*ghWebhookReciever, error) {
+	oauth2token, err := ioutil.ReadFile(cfg.authfile)
+	if err != nil {
+		return nil, err
+	}
+	ts := oauth2.StaticTokenSource(
+		&oauth2.Token{AccessToken: string(oauth2token)},
+	)
+	ctx := context.Background()
+	tc := oauth2.NewClient(ctx, ts)
+	return &ghWebhookReciever{
+		ghClient: github.NewClient(tc),
+		cfg:      cfg,
+	}, nil
+}
+
+// processAlert formats and posts the comment to github and returns nil if successful.
+func (g ghWebhookReciever) processAlert(ctx context.Context, msg *notify.WebhookMessage) error {
+
+	msgBody, err := formatIssueBody(msg)
+	if err != nil {
+		return err
+	}
+	issueComment := github.IssueComment{Body: &msgBody}
+
+	prNum, err := getTargetPR(msg)
+	if err != nil {
+		return err
+	}
+
+	_, _, err = g.ghClient.Issues.CreateComment(ctx,
+		g.getTargetOwner(msg), g.getTargetRepo(msg), prNum, &issueComment)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func serveWebhook(client *ghWebhookReciever) {
+	hl := ghWebhookHandler{client}
+	http.Handle("/hook", hl)
+	log.Printf("finished setting up gh client. starting amgh with %v/%v as defaults",
+		client.cfg.defaultOwner, client.cfg.defaultRepo)
+	log.Fatal(http.ListenAndServe(fmt.Sprintf(":%v", client.cfg.portNo), nil))
+}
+
+func main() {
```

Comment:
I would prefer to have the main func at the top to follow a more natural scrolling - start from the top and follow what is happening in the main func